src/eval/array.rs

```diff
@@ -276,23 +277,47 @@ impl Array {
         Ok(result)
     }
 
-    /// Return a sorted version of this array.
+    /// Return a sorted version of this array, optionally by a given key function.
     ///
-    /// Returns an error if two values could not be compared.
-    pub fn sorted(&self) -> StrResult<Self> {
+    /// Returns an error if two values could not be compared or if the key function (if given)
+    /// yields an error.
+    pub fn sorted(
+        &self,
+        vm: &mut Vm,
+        span: Span,
+        key: Option<Func>,
+    ) -> SourceResult<Self> {
         let mut result = Ok(());
         let mut vec = self.0.clone();
+        let mut key_ = |x: Value| match &key {
+            Some(f) => f.call_vm(vm, Args::new(f.span(), [x])),
+            None => Ok(x),
+        };
         vec.make_mut().sort_by(|a, b| {
-            a.partial_cmp(b).unwrap_or_else(|| {
-                if result.is_ok() {
-                    result = Err(eco_format!(
-                        "cannot order {} and {}",
-                        a.type_name(),
-                        b.type_name(),
-                    ));
+            // Until we get `try` blocks :)
+            match (key_(a.clone()), key_(b.clone())) {
+                (Ok(a), Ok(b)) => a.partial_cmp(&b).unwrap_or_else(|| {
+                    if result.is_ok() {
+                        result = Err(eco_format!(
+                            "cannot order {} and {}",
+                            a.type_name(),
+                            b.type_name(),
+                        ))
+                        .at(span);
+                    }
+                    Ordering::Equal
+                }),
+                (Err(mut e), _) | (_, Err(mut e)) => {
+                    if result.is_ok() {
+                        e.push(SourceError::new(
+                            span,
+                            "error while evaluating key for sorting",
+                        ));
```

Comment:
not sure whether we need this here. the error will probably overlap with the original error in the key closure.

Reply:
If we don't have this then the user has no idea that the key closure got evaluated due to the `.sorted` method

Reply:
would also be good to have one test for this

Reply:
but if you pass the closure to `sorted` then isn't it obvious that it will be called by `sorted`?

Reply:
The alternative is to use the `trace` functionality on errors (from `diag.rs`). This makes the blue underline in the web app similar to stack traces. Furthermore, it automatically removes unhelpful traces that surround the original error.

Reply:
> The alternative is to use the trace functionality on errors (from diag.rs). This makes the blue underline in the web app similar to stack traces. Furthermore, it automatically removes unhelpful traces that surround the original error.

Ah yeah, that sounds like what I wanted! I'll switch to that.

Reply:
Note that for your example, the trace wouldn't be shown because it would overlap with the error.

Reply:
But I honestly think that the second version is pretty clear.

Reply:
Yes, I agree, but what if we had:

```
#let my_key = x => x.length
#repr(#my_key("hello"))
#repr((-1, 0, 2).sorted(key: my_key))
```

Then we should have an additional message.

Reply:
Actually, I just realized that there will be an automatic trace for the method call, as shown below:
<img width="392" alt="example" src="https://user-images.githubusercontent.com/17899797/230572299-c842e7f8-e404-45ee-9011-9bc4d578bee1.png">
(map doesn't have any extra code for this)